File: library/src/main/java/me/proxer/library/api/list/MediaSearchEndpoint.java
package me.proxer.library.api.list;

import lombok.Setter;
import lombok.experimental.Accessors;
import me.proxer.library.api.PagingLimitEndpoint;
import me.proxer.library.api.ProxerCall;
import me.proxer.library.entity.list.MediaListEntry;
import me.proxer.library.enums.FskConstraint;
import me.proxer.library.enums.Language;
import me.proxer.library.enums.LengthBound;
import me.proxer.library.enums.MediaSearchSortCriteria;
import me.proxer.library.enums.MediaType;
import me.proxer.library.enums.TagRateFilter;
import me.proxer.library.enums.TagSpoilerFilter;
import me.proxer.library.util.ProxerUtils;

import javax.annotation.Nullable;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

/**
 * Search for all available media. Features various filter and sort options and uses paging.
 *
 * @author Desnoo
 */
@Accessors(fluent = true)
public final class MediaSearchEndpoint implements PagingLimitEndpoint<List<MediaListEntry>> {

    private static final String DELIMITER = "+";

    private final InternalApi internalApi;

    /**
     * Sets the name to search for.
     */
    @Nullable
    @Setter
    private String name;

    /**
     * Sets the language to filter by.
     */
    @Nullable
    @Setter
    private Language language;

    /**
     * Sets the type to load.
     */
    @Nullable
    @Setter
    private MediaType type;

    /**
     * Sets the criteria the list should be sorted by.
     */
    @Nullable
    @Setter
    private MediaSearchSortCriteria sort;

    /**
     * Sets the minimum/maximum episode count a entry must have to be included in the result.
     * You can specify if the count must be greater or smaller with the
     * {@link #lengthBound(LengthBound)} method.
     */
    @Nullable
    @Setter
    private Integer length;

    /**
     * To be used in conjunction with {@link #length(Integer)}. Sets if the episode count must
     * be greater or smaller than the specified value.
     */
    @Nullable
    @Setter
    private LengthBound lengthBound;

    /**
     * Sets the filter for the tags.
     */
    @Nullable
    @Setter
    private TagRateFilter tagRateFilter;

    /**
     * Sets the spoiler filter for the tags.
     */
    @Nullable
    @Setter
    private TagSpoilerFilter tagSpoilerFilter;

    /**
     * Sets the required fsk ratings.
     */
    @Nullable
    @Setter
    private EnumSet<FskConstraint> fskConstraints;

    /**
     * Sets if already finished media of the current user should be hidden from the result.
     */
    @Nullable
    @Setter
    private Boolean hideFinished;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter(onMethod = @__({@Override}))
    private Integer page;

    /**
     * {@inheritDoc}
     */
    @Nullable
    @Setter(onMethod = @__({@Override}))
    private Integer limit;

    @Nullable
    private String tags;

    @Nullable
    private String excludedTags;

    @Nullable
    private String genres;

    @Nullable
    private String excludedGenres;

    MediaSearchEndpoint(final InternalApi internalApi) {
        this.internalApi = internalApi;
    }

    /**
     * Sets the tag ids a entry must have to be included in the result.
     */
    public MediaSearchEndpoint tags(@Nullable final Set<String> ids) {
        if (ids != null) {
            this.tags = ProxerUtils.join(DELIMITER, ids);
        } else {
            this.tags = null;
        }

        return this;
    }

    /**
     * Sets the tag ids a entry must not have to be included in the result.
     */
    public MediaSearchEndpoint excludedTags(@Nullable final Set<String> excludedIds) {
        if (excludedIds != null) {
            this.excludedTags = ProxerUtils.join(DELIMITER, excludedIds);
        } else {
            this.excludedTags = null;
        }

        return this;
    }

    /**
     * Sets the genre tag ids a entry must have to be included in the result.
     */
    public MediaSearchEndpoint genres(@Nullable final Set<String> ids) {
        if (ids != null) {
            this.genres = ProxerUtils.join(DELIMITER, ids);
        } else {
            this.genres = null;
        }

        return this;
    }

    /**
     * Sets the genre tag ids a entry must not have to be included in the result.
     */
    public MediaSearchEndpoint excludedGenres(@Nullable final Set<String> excludedIds) {
        if (excludedIds != null) {
            this.excludedGenres = ProxerUtils.join(DELIMITER, excludedIds);
        } else {
            this.excludedGenres = null;
        }

        return this;
    }

    @Override
    public ProxerCall<List<MediaListEntry>> build() {
        final String joinedFskConstraints = fskConstraints == null ? null
                : ProxerUtils.joinEnums(DELIMITER, fskConstraints);

        return internalApi.mediaSearch(name, language, type, joinedFskConstraints, sort, length, lengthBound,
                tags, excludedTags, genres, excludedGenres, tagRateFilter, tagSpoilerFilter,
                hideFinished == null ? null : hideFinished ? 1 : 0, page, limit);
    }
}

File: library/src/main/kotlin/me/proxer/library/api/list/MediaSearchEndpoint.kt
package me.proxer.library.api.list

import me.proxer.library.ProxerCall
import me.proxer.library.api.PagingLimitEndpoint
import me.proxer.library.entity.list.MediaListEntry
import me.proxer.library.enums.FskConstraint
import me.proxer.library.enums.Language
import me.proxer.library.enums.LengthBound
import me.proxer.library.enums.MediaSearchSortCriteria
import me.proxer.library.enums.MediaType
import me.proxer.library.enums.TagRateFilter
import me.proxer.library.enums.TagSpoilerFilter
import me.proxer.library.internal.util.toIntOrNull
import me.proxer.library.util.ProxerUtils

/**
 * Search for all available media. Features various filter and sort options and uses paging.
 *
 * @author Desnoo
 */
class MediaSearchEndpoint internal constructor(
    private val internalApi: InternalApi
) : PagingLimitEndpoint<List<MediaListEntry>> {

    private companion object {
        private const val DELIMITER = "+"
    }

    private var page: Int? = null
    private var limit: Int? = null

    private var name: String? = null
    private var language: Language? = null
    private var type: MediaType? = null
    private var sort: MediaSearchSortCriteria? = null
    private var length: Int? = null
    private var lengthBound: LengthBound? = null
    private var tagRateFilter: TagRateFilter? = null
    private var tagSpoilerFilter: TagSpoilerFilter? = null
    private var fskConstraints: Set<FskConstraint>? = null
    private var hideFinished: Boolean? = null
    private var tags: Set<String>? = null
    private var excludedTags: Set<String>? = null
    private var genres: Set<String>? = null
    private var excludedGenres: Set<String>? = null

    override fun page(page: Int?) = this.apply { this.page = page }
    override fun limit(limit: Int?) = this.apply { this.limit = limit }

    /**
     * Sets the name to search for.
     */
    fun name(name: String?) = this.apply { this.name = name }

    /**
     * Sets the language to filter by.
     */
    fun language(language: Language?) = this.apply { this.language = language }

    /**
     * Sets the type to load.
     */
    fun type(type: MediaType?) = this.apply { this.type = type }

    /**
     * Sets the criteria the list should be sorted by.
     */
    fun sort(sort: MediaSearchSortCriteria?) = this.apply { this.sort = sort }

    /**
     * Sets the minimum/maximum episode count a entry must have to be included in the result.
     * You can specify if the count must be greater or smaller with the [lengthBound] method.
     */
    fun length(length: Int?) = this.apply { this.length = length }

    /**
     * To be used in conjunction with [length]. Sets if the episode count must
     * be greater or smaller than the specified value.
     */
    fun lengthBound(lengthBound: LengthBound?) = this.apply { this.lengthBound = lengthBound }

    /**
     * Sets the filter for the tags.
     */
    fun tagRateFilter(tagRateFilter: TagRateFilter?) = this.apply { this.tagRateFilter = tagRateFilter }

    /**
     * Sets the spoiler filter for the tags.
     */
    fun tagSpoilerFilter(tagSpoilerFilter: TagSpoilerFilter?) = this.apply { this.tagSpoilerFilter = tagSpoilerFilter }

    /**
     * Sets the required fsk ratings.
     */
    fun fskConstraints(fskConstraints: Set<FskConstraint>?) = this.apply { this.fskConstraints = fskConstraints }

    /**
     * Sets if already finished media of the current user should be hidden from the result.
     */
    fun hideFinished(hideFinished: Boolean?) = this.apply { this.hideFinished = hideFinished }

    /**
     * Sets the tag ids a entry must have to be included in the result.
     */
    fun tags(ids: Set<String>?) = this.apply { this.tags = ids }

    /**
     * Sets the tag ids a entry must not have to be included in the result.
     */
    fun excludedTags(excludedIds: Set<String>?) = this.apply { this.excludedTags = excludedIds }

    /**
     * Sets the genre tag ids a entry must have to be included in the result.
     */
    fun genres(ids: Set<String>?) = this.apply { this.genres = ids }

    /**
     * Sets the genre tag ids a entry must not have to be included in the result.
     */
    fun excludedGenres(excludedIds: Set<String>?) = this.apply { this.excludedGenres = excludedIds }

    override fun build(): ProxerCall<List<MediaListEntry>> {
        val joinedFskConstraints = fskConstraints
            ?.map { ProxerUtils.getSafeApiEnumName(it) }
            ?.joinToString(DELIMITER)

        val joinedTags = tags?.joinToString(DELIMITER)
        val joinedExcludedTags = excludedTags?.joinToString(DELIMITER)
        val joinedGenres = genres?.joinToString(DELIMITER)
        val joinedExcludedGenres = excludedGenres?.joinToString(DELIMITER)

        return internalApi.mediaSearch(
            name, language, type, joinedFskConstraints, sort, length, lengthBound,
            joinedTags, joinedExcludedTags, joinedGenres, joinedExcludedGenres,
            tagRateFilter, tagSpoilerFilter, hideFinished.toIntOrNull(), page, limit
        )
    }
}
